File: ios/Stopwatch/infra/presentation/view_models/HomeViewModel.swift
import Foundation

struct HomeState {
    let status: Status
    let time: ViewTime
    let laps: [ViewLap]
    let showLaps: Bool
    let showSeeAll: Bool
    
    init(
        status: Status = .initial,
        time: ViewTime = ViewTime(),
        laps: [ViewLap] = [],
        showLaps: Bool = false,
        showSeeAll: Bool = false
    ) {
        self.status = status
        self.time = time
        self.laps = laps
        self.showLaps = showLaps
        self.showSeeAll = showSeeAll
    }
}

enum HomeAction {
    case start, pause, resume, reset, lap, seeAll
}

protocol HomeViewModel: ObservableObject {
    var state: HomeState { get }

    func handleAction(_ action: HomeAction)
}

File: android/app/src/main/java/com/rafaeltmbr/stopwatch/platform/presentation/view_models/HomeViewModel.kt
package com.rafaeltmbr.stopwatch.platform.presentation.view_models

import com.rafaeltmbr.stopwatch.core.entities.Status
import com.rafaeltmbr.stopwatch.platform.presentation.entities.ViewLap
import com.rafaeltmbr.stopwatch.platform.presentation.entities.ViewTime
import kotlinx.coroutines.flow.StateFlow


interface HomeViewModel {
    data class State(
        val status: Status = Status.INITIAL,
        val time: ViewTime = ViewTime(
            minutes = listOf("0", "0"),
            seconds = listOf("0", "0"),
            fraction = listOf("0", "0")
        ),
        val laps: List<ViewLap> = emptyList(),
        val showLapsSection: Boolean = false,
        val showSeeAllLaps: Boolean = false
    )

    sealed class Action {
        data object Start : Action()
        data object Pause : Action()
        data object Resume : Action()
        data object Reset : Action()
        data object Lap : Action()
        data object SeeAll : Action()
    }

    val state: StateFlow<State>
    fun handleAction(action: Action)
}
